neural_lam/weather_dataset.py

```diff
@@ -294,46 +431,115 @@ def _slice_forcing_time(self, da_forcing, idx, n_steps: int):
+            # Create window dimension for forcing data to stack later
+            window_size = num_past_steps + num_future_steps + 1
+            da_forcing_boundary_windowed = da_forcing_boundary.rolling(
+                time=window_size, center=False
+            ).construct(window_dim="window")
+            da_forcing_boundary_matched = da_forcing_boundary_windowed.isel(
+                time=idx_min[init_steps:]
+            )
 
-                da_sliced = da_sliced.rename({"time": "window"})
+        # Add time difference as a new coordinate to concatenate to the
+        # forcing features later as temporal embedding
+        da_forcing_boundary_matched["time_diff_steps"] = (
+            ("time", "window"),
+            time_diff_steps,
+        )
 
-                # Assign the 'window' coordinate to be relative positions
-                da_sliced = da_sliced.assign_coords(
-                    window=np.arange(len(da_sliced.window))
-                )
+        return da_state_sliced, da_forcing_boundary_matched
 
-                # Add a 'time' dimension to keep track of steps using actual
-                # time coordinates
-                current_time = da_forcing.time[offset + step]
-                da_sliced = da_sliced.expand_dims(
-                    dim={"time": [current_time.values]}
-                )
+    def _process_windowed_data(self, da_windowed, da_state, da_target_times):
+        """Helper function to process windowed data. This function stacks the
+        'forcing_feature' and 'window' dimensions and adds the time step
+        differences to the existing features as a temporal embedding.
 
-                da_list.append(da_sliced)
+        Parameters
+        ----------
+        da_windowed : xr.DataArray
+            The windowed data to process. Can be `None` if no data is provided.
+        da_state : xr.DataArray
+            The state dataarray.
+        da_target_times : xr.DataArray
+            The target times.
 
-            # Concatenate the list of DataArrays along the 'time' dimension
-            da_concat = xr.concat(da_list, dim="time")
+        Returns
+        -------
+        da_windowed : xr.DataArray
+            The processed windowed data. If `da_windowed` is `None`, an empty
+            DataArray with the correct dimensions and coordinates is returned.
 
-        return da_concat
+        """
+        stacked_dim = "forcing_feature_windowed"
+        if da_windowed is not None:
+            # Stack the 'feature' and 'window' dimensions and add the
+            # time step differences to the existing features as a temporal
+            # embedding
+            da_windowed = da_windowed.stack(
+                {stacked_dim: ("forcing_feature", "window")}
+            )
+            da_windowed = xr.concat(
+                [da_windowed, da_windowed.time_diff_steps],
+                dim="forcing_feature_windowed",
+            )
```

Comment:
How will the concatenation of these time deltas work with e.g. `BaseDatastore.get_vars_names(category="forcing")`? I guess they will not be included there. But that also does not reflect the forcing windowing, so maybe one should not make the assumption that what is returned there maps to the indices of the forcing tensor returned from the dataset? That would require the datastore to have knowledge of the WeatherDataset config, which we probably don't want :thinking: 

I fully agree that we need these time-delta features, but it's tricky how to handle them. We've been talking about doing some form of sinusoidal embeddings of them later in the model. Would it even make sense to return these separately? Since if we  want to do something special (embedding) with these features we would otherwise have to extract them from the boundary forcing tensor (assuming that they are the last features). If we would return them separately we should maybe create some form of custom batch-object, as it would start to be a little unstructured to have this tuple of 7 objects.

Reply:
Yes, BaseDatastore.get_vars_names(category="forcing") will not return the windowed_features + temporal_embeddings. This is important for saving model predictions in https://github.com/mllam/neural-lam/issues/89, too.  We don't need them in the model output, but want to keep the original features exactly in order.

- ordering the forcing and boundary return, to always have the temporal embedding as the last feature, sound like an easy option for now
- I am still a bit confused how exactly the temporal embedding will be used in the model. The temporal embedding is the same for all `grid_points, time` and maps to 'windowed_forcing_features` based on the window-size. Can one just leave these additional temporal embeddings as additional input feature channels (sin(embedding)), or should we directly _merge_ them with the forcing_feature in some sort of batch-norm in ar_model? Is there a world where we can already merge the embedding and forcing_feature in weatherDataset?

Reply:
> Can one just leave these additional temporal embeddings as additional input feature channels (sin(embedding))

I guess we should leave it open to different model implementations what to do with these. What I am thinking of first is to just do the sin(embedding) and feed them as additional input feature channels. Then we could expand on that later as needed.

> or should we directly merge them with the forcing_feature in some sort of batch-norm in ar_model?

That's another option, to map from these features to the mean and std-dev used in LayerNorms (akin to how time-embeddings are often used in diffusion-models and others). It's a bit trickier to implement, and might not really be neccesary here, but still potentially interesting.

> Is there a world where we can already merge the embedding and windowed_forcing_features in weatherDataset?

I don't think that would be possible (considering the LayerNorm setup discussed above). As the use of theses features in the LayerNorms involves trainable parameters, and they act on intermediate representations in the model, this would have to be done on the model side of things.

I suppose we could do the sin(embedding) already in WeatherDataset, if that does not involve trainable parameters. But I don't think we want to, as it should be better to do those computations on GPU (similar to how we want to move standardization to GPU).

Reply:
okay thanks for the explanations. While not totally relevent here, it's still good to have the bigger picture in mind. so for now I just focus un returning all temporal embedding features as the last channels in `forcing` or `boundary`

Reply:
Okay I fixed this now in https://github.com/mllam/neural-lam/pull/90/commits/cfe1e278ae16a4ec8e19dc3f2db79e976e8014d6 
the temporal embedding has size `window` and is not repeated for each feature.
the embedding is concatenated to the existing `forcing_feature_windowed` at the very end